File: sypet/all-in-one/src/main/kotlin/edu/cmu/sypet/learner/Learner.kt
package edu.cmu.sypet.learner

import edu.cmu.sypet.learner.sketchSolver.Program
import edu.cmu.sypet.learner.sketchSolver.SketchSolver
import edu.cmu.sypet.learner.sketcher.Sketch
import edu.cmu.sypet.learner.sketcher.Sketcher

data class Query(val sketch: Sketch)

sealed class Response
object Yes : Response()
object No : Response()

data class Spec(val testCases: Collection<TestCase>)

interface TestCase

data class Learner(val sketcher: Sketcher, val sketchSolver: SketchSolver) {
    private var done: Boolean = false

    fun start(spec: Spec): Query {
        TODO()
    }

    fun learn(response: Response): Query {
        TODO()
    }

    fun isDone(): Boolean {
        TODO()
    }

    fun stop(): Program? {
        TODO()
    }
}

File: src/CMU.SyPet.Originals/Internal/Learner.cs
using System.Linq;
using System.Collections.Generic;
using CMU.Commons.Types;
using CMU.SyPet.Synthesis;

namespace CMU.SyPet.Originals.Internals
{
    /// <summary>The original SyPet learner based on Petri net reachability.</summary>
    internal sealed class Learner :
        SketchBasedLearner<Spec, Program, Program, Response, Sketch, Unit, Unit>
    {
        public Learner(
            ISketcher<Spec, Sketch, Unit> sketcher,
            ISketchSolver<Sketch, Program, Unit> sketchSolver,
            Spec spec
        ) : base(sketcher, sketchSolver, spec) { }

        protected override Program CreateQuery(Program program) => program;

        protected override Program? CreateResult(Response response) => response.Program;

        // Our sketcher is dumb and does not know how to take a hint.
        protected override Unit CreateSketcherHint(Response response) => new Unit();

        // Our sketch solver is dumb and does not know how to take a hint.
        protected override Unit CreateSketchSolverHint(Response response) => new Unit();

        protected override bool Finished(Response response) => response.IsCorrectProgram;

        protected override IEnumerable<Sketch> SelectSketches(IEnumerable<Sketch> sketches) =>
            sketches;

        protected override Program SelectProgram(IEnumerable<Program> programs) =>
            programs.First();
    }
}
